### ISSUE TITLE
Allow impersonating a Kubernetes user/group with istioctl

### ISSUE THREAD

**Issue by jonmoter** [istio/istio#22365]
**Describe the feature request**

Request is to add `--as` and `--as-group` flags to `istioctl`, similar to what exists with `kubectl`. That way I can leverage user impersonation.

Kubernetes docs: https://kubernetes.io/docs/reference/access-authn-authz/authentication/#user-impersonation

I have my Kubernetes cluster configured so the default user I authenticate with has read-only permissions. That way I'm less likely to accidentally delete or break something. But my user has the ability to impersonate another user with greater permissions.

```
$ pod_name=kube-shell-cnl7r
$ kubectl delete pod ${pod_name}
Error from server (Forbidden): pods "kube-shell-cnl7r" is forbidden: User "[USER_NAME_REDACTED]" cannot delete resource "pods" in API group "" in the namespace "mesh-enabled"
$ kubectl --as=admin --as-group=system:masters delete pod ${pod_name}
pod "kube-shell-cnl7r" deleted
```

That means I can't run istioctl as my user, and can't impersonate another user.

```
$ istioctl version
Error: 13 errors occurred:
	* error execing into istio-citadel-77d6775984-bmsml citadel container: pods "istio-citadel-77d6775984-bmsml" is forbidden: User "[USER_NAME_REDACTED]" cannot create resource "pods/exec" in API group "" in the namespace "istio-system"
	* error execing into istio-citadel-77d6775984-j5zwp citadel container: pods "istio-citadel-77d6775984-j5zwp" is forbidden: User "[USER_NAME_REDACTED]" cannot create resource "pods/exec" in API group "" in the namespace "istio-system"
```

**Describe alternatives you've considered**

As far as I know, there's no way to modify my `~/.kube/config` file to specify I want to  impersonate a user.

**Additional context**

Using version 1.5.0 of istioctl.

**Comment by howardjohn** [comment]
related: https://github.com/istio/istio/issues/19927

**Comment by esnible** [comment]
@jonmoter This looks like a good idea but I don't have experience using this with _kubectl_.  I have put it on the UX working group agenda to discuss April 7th so that I can get feedback from others.  @istio/wg-security-maintainers please comment if there are security concerns.  (I don't believe there are).

**Comment by jonmoter** [comment]
FYI, I found a workaround, in case anyone else needs similar behavior. You can put impersonation information into your kubeconfig file, so the `users` section would have an entry like:

```
users:
- name: myuser-as-admin
  user:
    # Normal credentials here, could be client certs, exec, whatever
    username: myuser
    password: password123

    # Impersonation information below
    as: admin
    as-groups:
      - system:masters
    as-user-extra:
      reason:
        - istioctl
```

Then create a context that refers to that user, and then you can run `istioctl --context=mycluster-as-admin ...`

I still think it would be nice to have them as flags for istioctl (especially if you leverage `genericclioptions` as mentioned in #19927), but the above workaround unblocked me.

**Comment by esnible** [comment]
@istio/wg-security-maintainers I never use impersonation.  I usually run with cluster admin permissions.  Is this something that helps a user of a shared cluster troubleshoot issues other users are having with _istioctl_?  Should we prioritize this work?

@linsun @rvennam Do we have customers who are getting issues that are hard to understand because the user getting the problem and the use capable of fixing it are two different people?

**Comment by jonmoter** [comment]
Like I said above, once I discovered that you can configure impersonation in a kubeconfig file, this was no longer an issue for me, so I don't think adding impersonation as a CLI flag is necessary.  As the original author of the issue, I'm fine with closing this if you'd like.

> Do we have customers who are getting issues that are hard to understand because the user getting the problem and the use capable of fixing it are two different people?

FWIW, istioctl has a broad mix of functionality. Some commands should only ever be done by a cluster admin, like `istioctl operator ...` or `istioctl register`.  But there are some other things that I could see being useful for users running workloads on a cluster, like `istioctl dashboard ...` or `istioctl experimental describe pod`.

At Zendesk, we have a central team managing the k8s clusters and Istio control plane, but we expect the teams owning any given service/app to manage the k8s resources related to their service. We're still onboarding teams and rolling out Istio, so it remains to be seen how much individual teams will need to use istioctl.  I'll share with the community what we find, and what (if anything) turns out to be a real blocker.

**Comment by howardjohn** [comment]
not stale
